include/nanogui/widget.h

```diff
@@ -334,7 +413,66 @@ class NANOGUI_EXPORT Widget : public Object {
      * \endrst
      */
     float mIconExtraScale;
+
     Cursor mCursor;
+
+private:
+    /* Font related utilities.  To use in derived classes, call the getter.
+     * E.g. for `mFont`, call `font()` and `mTooltipFont` call `tooltipFont()`.
+     * To set it, call the setter.  E.g. set `mFont` with `setFont` and
+     * `mTooltipFont` with `setTooltipFont`.
+     */
+
+    /**
+     * When \ref setTheme is called, this method configures both \ref mFont and
+     * \ref mTooltipFont.
+     */
+    void setDefaultFonts();
+
+    /**
+     * Used to set the font size of a widget explicitly.  The initial value is
+     * ``-1``, and a negative number indicates that the theme's font size should
+     * be used instead.  Specifically, \ref fontSize will return
+     * \ref Theme::mStandardFontSize when ``mFontSize < 0``.
+     */
+    float mFontSize;
+
+    /// The current font face being used to draw text (if this Widget displays text).
+    std::string mFont;
+    /**
+     * When the font is specified either via the constructor or \ref setFont,
+     * this is set to ``true`` to indicate that \ref setTheme should not
+     * overwrite the user-specified default.  Otherwise, when \ref setTheme is
+     * called, the theme's default window font will be used.
+     */
+    bool mFontExplicit = false;
```

Comment:
Hi @svenevs,

the amount of storage and logic dedicated to keep track of fonts in the ``Widget`` class has me somewhat concerned. I feel like it's best if there is one more iteration that simplifies this as much as possible.

For starters, why do we need ``mFontExplicit``? I assume that the font will just be an empty string initially. ``setFont`` could be implemented so that it only replaces the font string in ``mFont`` when it is equal to the empty string (or the corresponding font string from the previous theme, which I suppose is quite rare usage and shouldn't drive complexity too much here).

Reply:
Yes, there has to be a better way to approach this.  The tension is really just that there is no certainty about when `setTheme` is called (e.g. before or after `new Label(parent, "text", "sans-bold")`), so these really only exist to ensure that `setTheme` does not overwrite user-specified font values.

So it sounds like the goal here would be to merge `mFontExplicit` to be implicit when `mFont == ""`?.  So (ignoring bold stuff), `font()` could change to be

```cpp
if (mFont != "")
    return mFont;
return mTheme->defaultFont();
```

That alleviates the concerns about theme setting overwriting user-specified values.  I don't see a solution for default vs bold other than a "magic" value.  E.g.

```cpp
if (mFont != "") {
    if (mFont == "THEME_BOLD")
        return mTheme->defaultBoldFont();
    else
        return mFont;
}
return mTheme->defaultFont();
```

I was trying to avoid stuff like this because now we're forcing a virtual method into the rendering loop for all widgets using fonts, when most users likely won't be using custom themes in the first place.

I will need to think more carefully about how to solve this.

Reply:
@wjakob I've thought about this more, and I don't think avoiding virtual methods can be done here.  Before implementing, do you think this is a good approach?

1. `mFont` is the only `Widget` variable.  It remains `private` necessitating users to call `font()`.
2. `Widget::font()` implementation becomes

    ```cpp
    if (mFont != "")
        return mFont;
    return defaultFont();
    ```
3. Remove font related methods in `Theme` in favor of three instance variables (also makes customization even easier): `Theme::mDefaultFont`, `Theme::mDefaultBoldFont`, and `Theme::mDefaultIconFont`.
4. Introduce a virtual `Widget::defaultFont()`, which subclasses override if they want to specify a default font of bold.

    ```cpp
    virtual std::string defaultFont() {
        if (mTheme)
            return mTheme->mDefaultFont;
        return "";
    }

    The only time the empty string will be returned is if somebody explicitly created an orphan widget.  But as far as I understand, it is assumed that `mTheme` will exist when rendering takes place.  So do we even want the check?

Reply:
+1 for this approach by the way.